src/v/wasm/allocator.h

```diff
@@ -72,13 +72,19 @@ class heap_allocator {
 
     /**
      * Allocate heap memory by taking a memory instance from the pool.
+     *
+     * Memory returned from this method will be zero-filled.
      */
     std::optional<heap_memory> allocate(request);
 
     /**
      * Deallocate heap memory by returing a memory instance to the pool.
+     *
+     * used_amount is to zero out the used memory from the heap, this is
+     * required so that if only a portion of a large memory space was used we
+     * don't have to touch all the bytes.
      */
-    void deallocate(heap_memory);
+    void deallocate(heap_memory, size_t used_amount);
```

Comment:
This amount is user configurable and in theory could be 10s of MB, I guess that is too expensive to memset in a single go? Do we need to do this in an async friendly way?

Reply:
Like chunk-wise with yields inerspersed or something?

Reply:
yeah

Reply:
Does N have a concrete upper bound?

Reply:
4GiB, probably in practice it's much lower (the default is 2MiB), but for example golang (not tinygo) is requiring 50MB in my tests for a non trivial program (seems like a lot but 🤷)

Reply:
I have an idea to do this in a fairly lightweight way, but I will do in a followup PR, not this one.

Reply:
i would assume that as long as memset(0 is on a contiguous region it is very fast. i dunno tho what the limit is where it becomes costly enough to be a concern.

Reply:
It's also page aligned, that should help